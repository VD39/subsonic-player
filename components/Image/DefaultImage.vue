<script setup lang="ts">
defineProps<{
  icon: string;
}>();
</script>

<template>
  <div :class="$style.defaultImage">
    <component :is="icon" ref="iconComponent" :class="$style.icon" />
  </div>
</template>

<style module>
.defaultImage {
  position: relative;
  display: block;
  width: 0;
  height: 0;
  padding-top: 100%;
  padding-left: 100%;
  background-color: var(--secondary-background-color);
  border-radius: var(--border-radius-medium);
}

.icon {
  position: absolute;
  inset: 50% auto auto 50%;
  width: 100%;
  max-width: 45px;
  height: 100%;
  transform: translate(-50%, -50%);
}
</style>
