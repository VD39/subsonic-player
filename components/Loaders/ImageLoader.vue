<template>
  <span :class="$style.loader">
    <span class="visually-hidden">Loading image</span>
  </span>
</template>

<style module>
.loader {
  display: block;
  overflow: hidden;

  &::after {
    position: absolute;
    inset: 0 0 auto;
    z-index: 1;
    height: 100%;
    content: '';
    background: linear-gradient(
      90deg,
      transparent,
      var(--border-color),
      transparent
    );
    transform: translateX(-100%);
    animation: image-loader 1.3s infinite;
  }
}

@keyframes image-loader {
  from {
    transform: translateX(-100%);
  }

  to {
    transform: translateX(100%);
  }
}
</style>
