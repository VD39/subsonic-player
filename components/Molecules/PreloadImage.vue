<script setup lang="ts">
import IconImage from '@/components/Atoms/IconImage.vue';

const props = withDefaults(
  defineProps<{
    alt?: string;
    image: Image;
  }>(),
  {
    alt: 'Image',
  },
);

const { getImageUrl } = useAPI();

const loading = ref(true);

function onImageLoad() {
  loading.value = false;
}

const imageSrc = computed(() => {
  if (Object.values(IMAGE_DEFAULT_BY_TYPE).includes(props.image as Icon)) {
    return null;
  }

  if (isUrl(props.image)) {
    return props.image;
  }

  return getImageUrl(props.image);
});
</script>

<template>
  <div :class="['centerAll', $style.preloadImage]">
    <template v-if="imageSrc">
      <span v-show="loading" ref="imageLoader" :class="$style.imageLoader">
        <span class="visuallyHidden">Loading image</span>
      </span>

      <ClientOnly>
        <img
          v-show="!loading"
          ref="img"
          :alt="alt"
          :class="$style.image"
          :src="imageSrc"
          @load="onImageLoad"
        />
      </ClientOnly>
    </template>

    <IconImage v-else :icon="image as Icon" />
  </div>
</template>

<style module>
.preloadImage {
  --preload-opacity: 1;

  position: relative;
  flex-shrink: 0;
  aspect-ratio: 1;
  overflow: hidden;
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-medium);
  box-shadow: var(--box-shadow-medium);
  opacity: var(--preload-opacity);

  @media (hover: hover) {
    &:hover {
      --preload-opacity: 0.85;
    }
  }
}

.image {
  height: var(--width-height-100);
  object-fit: cover;
}

.imageLoader {
  display: block;
  overflow: hidden;

  &::after {
    position: absolute;
    inset: 0;
    z-index: 1;
    height: var(--width-height-100);
    content: '';
    background: linear-gradient(
      90deg,
      transparent,
      var(--border-color),
      transparent
    );
    transform: translateX(-100%);
    animation: image-loader 1.5s infinite;
  }
}

@keyframes image-loader {
  from {
    transform: translateX(-100%);
  }

  to {
    transform: translateX(100%);
  }
}
</style>
