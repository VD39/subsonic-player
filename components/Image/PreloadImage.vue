<script setup lang="ts">
import ImageLoader from '@/components/Loaders/ImageLoader.vue';
import DefaultImage from './DefaultImage.vue';

const props = withDefaults(
  defineProps<{
    image: string;
    alt?: string;
  }>(),
  {
    alt: 'Image',
  },
);

const { getImageUrl } = useAPI();

const isLoading = ref(true);

function onImageLoad() {
  isLoading.value = false;
}

const imageSrc = computed(() => {
  if (DEFAULT_IMAGE_ICONS.includes(props.image)) {
    return null;
  }

  if (isUrl(props.image)) {
    return props.image;
  }

  return getImageUrl(props.image);
});
</script>

<template>
  <div :class="$style.preloadImage">
    <template v-if="imageSrc">
      <ImageLoader v-show="isLoading" />

      <img
        v-show="!isLoading"
        ref="img"
        :class="$style.image"
        :src="imageSrc"
        :alt="alt"
        @load="onImageLoad"
      />
    </template>
    <DefaultImage v-else :icon="image" />
  </div>
</template>

<style module>
.preloadImage {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  aspect-ratio: 1;
  overflow: hidden;
  border-radius: var(--border-radius-small);
  box-shadow: var(--box-shadow-medium);
}

.image {
  height: 100%;
  object-fit: cover;
}
</style>
