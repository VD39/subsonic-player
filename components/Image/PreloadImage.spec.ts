import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import ImageLoader from '@/components/Loaders/ImageLoader.vue';
import DefaultImage from './DefaultImage.vue';
import PreloadImage from './PreloadImage.vue';

function factory(props = {}) {
  return mount(PreloadImage, {
    props: {
      image: 'imageId',
      ...props,
    },
    attachTo: document.body,
  });
}

describe('Default', () => {
  let wrapper: VueWrapper;

  beforeAll(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when image prop is set to a none DEFAULT_IMAGE_ICONS icon', () => {
    describe('when image prop is not a URL', () => {
      it('does not show the DefaultImage component', () => {
        expect(wrapper.findComponent(DefaultImage).exists()).toBe(false);
      });

      it('shows the img element', () => {
        expect(wrapper.find({ ref: 'img' }).exists()).toBe(true);
      });
    });

    describe('when image prop is a URL', () => {
      beforeAll(() => {
        wrapper = factory({
          image: 'https://test.com',
        });
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('does not show the DefaultImage component', () => {
        expect(wrapper.findComponent(DefaultImage).exists()).toBe(false);
      });

      it('shows the img element', () => {
        expect(wrapper.find({ ref: 'img' }).exists()).toBe(true);
      });
    });

    describe('when alt prop is not set', () => {
      it('adds the default alt attribute', () => {
        expect(wrapper.find({ ref: 'img' }).attributes('alt')).toBe('Image');
      });
    });

    describe('when alt prop is set to true', () => {
      beforeAll(() => {
        wrapper = factory({
          alt: 'Alt text for image',
        });
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('adds the correct alt attribute', () => {
        expect(wrapper.find({ ref: 'img' }).attributes('alt')).toBe(
          'Alt text for image',
        );
      });
    });

    describe('when image is loading', () => {
      it('shows the placeholder', () => {
        expect(wrapper.findComponent(ImageLoader).isVisible()).toBe(true);
      });

      it('does not show the image', () => {
        expect(wrapper.find({ ref: 'img' }).isVisible()).toBe(false);
      });
    });

    describe('when image has finished loading', () => {
      beforeAll(async () => {
        await wrapper.find({ ref: 'img' }).trigger('load');
        await wrapper.vm.$nextTick();
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('does not show the placeholder', () => {
        expect(wrapper.findComponent(ImageLoader).isVisible()).toBe(false);
      });

      it('shows the image', () => {
        expect(wrapper.find({ ref: 'img' }).isVisible()).toBe(true);
      });
    });
  });

  describe.each([...DEFAULT_IMAGE_ICONS])(
    'when image prop is set to %s',
    (image) => {
      beforeAll(() => {
        wrapper = factory({
          image,
        });
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('shows the DefaultImage component', () => {
        expect(wrapper.findComponent(DefaultImage).exists()).toBe(true);
      });

      it('does not show the img element', () => {
        expect(wrapper.find({ ref: 'img' }).exists()).toBe(false);
      });
    },
  );
});
