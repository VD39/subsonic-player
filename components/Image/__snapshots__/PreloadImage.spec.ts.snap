// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Default > matches the snapshot 1`] = `"<div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="imageId" alt="Image" style="display: none;"></div>"`;

exports[`Default > when image prop is set to PhApplePodcastsLogo > matches the snapshot 1`] = `
"<div class="preloadImage">
  <div class="defaultImage">
    <phapplepodcastslogo class="icon"></phapplepodcastslogo>
  </div>
</div>"
`;

exports[`Default > when image prop is set to PhMusicNotes > matches the snapshot 1`] = `
"<div class="preloadImage">
  <div class="defaultImage">
    <phmusicnotes class="icon"></phmusicnotes>
  </div>
</div>"
`;

exports[`Default > when image prop is set to PhPlaylist > matches the snapshot 1`] = `
"<div class="preloadImage">
  <div class="defaultImage">
    <phplaylist class="icon"></phplaylist>
  </div>
</div>"
`;

exports[`Default > when image prop is set to PhRadio > matches the snapshot 1`] = `
"<div class="preloadImage">
  <div class="defaultImage">
    <phradio class="icon"></phradio>
  </div>
</div>"
`;

exports[`Default > when image prop is set to PhUsersThree > matches the snapshot 1`] = `
"<div class="preloadImage">
  <div class="defaultImage">
    <phusersthree class="icon"></phusersthree>
  </div>
</div>"
`;

exports[`Default > when image prop is set to PhVinylRecord > matches the snapshot 1`] = `
"<div class="preloadImage">
  <div class="defaultImage">
    <phvinylrecord class="icon"></phvinylrecord>
  </div>
</div>"
`;

exports[`Default > when image prop is set to a none DEFAULT_IMAGE_ICONS icon > when alt prop is set to true > matches the snapshot 1`] = `"<div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="imageId" alt="Alt text for image" style="display: none;"></div>"`;

exports[`Default > when image prop is set to a none DEFAULT_IMAGE_ICONS icon > when image has finished loading > matches the snapshot 1`] = `"<div class="preloadImage"><span class="loader" style="display: none;"><span class="visually-hidden">Loading image</span></span><img class="image" src="imageId" alt="Alt text for image"></div>"`;

exports[`Default > when image prop is set to a none DEFAULT_IMAGE_ICONS icon > when image prop is a URL > matches the snapshot 1`] = `"<div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="https://test.com" alt="Image" style="display: none;"></div>"`;
